xlml/utils/tpu.py

```diff
@@ -372,3 +386,38 @@ def clean_up_idle_queued_resources(project_name: str, zones: Iterable[str]) -> N
       ):
         logging.info(f'Deleting {qr.name} in {state.name} status.')
         client.delete_queued_resource(name=qr.name)
+
+
+@task
+def clean_up_idle_nodes(project_name: str, zones: Iterable[str]) -> None:
+  """Clean up TPU nodes that are expired.
+
+  Args:
+   project_name: The project of resources.
+   zones: Available zones to clean up for the project.
+  """
+  creds, _ = google.auth.default()
+  client = tpu_api.TpuClient(credentials=creds)
+
+  logging.info(f'Cleaning up nodes in project {project_name}.')
+  for zone in zones:
+    logging.info(f'Checking in zone {zone.value}.')
+    parent = f'projects/{project_name}/locations/{zone.value}'
+    request = tpu_api.types.ListNodesRequest(parent=parent)
+    responses = client.list_nodes(request)
+
+    for node in responses:
+      create_time = node.create_time
+      current_time = datetime.datetime.now(datetime.timezone.utc)
+      ttl = int(node.labels[TTL]) if node.labels[TTL] else 0
```

Comment:
What is `node.labels[TTL]` for nodes without a TTL? `None`?

Reply:
I think it would be None, as this [labels block](https://screenshot.googleplex.com/8uvLmWPvmxWVTp8) won't be in the response.

Reply:
Correction, the labels will return empty dictionary if not exists ([here](https://screenshot.googleplex.com/9Jb9VAvZ9WBv5ND)).